### PR TITLE
AEX-9 Tipping Adjustments

### PR DESCRIPTION
**Pull request by thepiwo**
just a draft implementation for now, will work the coming days to polish and test the changes

### CODE REVIEW

File: logic/tokenCacheLogic.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+module.exports = class TokenCacheLogic {
+
+  constructor() {
+    this.init();
+  }
+
+  async init() {
+    await aeternity.init();
+    await cache.init(aeternity, () => {});
```
Comment (kenodressel) [inline logic/tokenCacheLogic.js:13]: do we need to init the cache twice?

File: test/tokenCache.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+let server = require('../server');
+
+chai.use(chaiHttp);
+
+describe('Token Cache', () => {
+  describe('API', () => {
+    it('it should GET token info', function (done) {
+      chai.request(server).get('/tokenCache/tokenInfo').end((err, res) => {
+        res.should.have.status(200);
+        res.body.should.be.a('object');
```
Comment (kenodressel) [inline test/tokenCache.js:12]: can we check on some properties here? to also verify that there is a proper response

File: test/tokenCache.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+      chai.request(server).post('/tokenCache/addToken')
+        .send({address: "ct_2DQ1vdJdiaNVgh2vUbTTpkPRiT9e2GSx1NxyU7JM9avWqj6dVf"}).end((err, res) => {
+        res.should.have.status(200);
+        res.text.should.be.equal('OK');
+        done();
+      });
+    });
+
+    it('it should GET token info', function (done) {
+      chai.request(server).get('/tokenCache/balances?address=' + process.env.PUBLIC_KEY).end((err, res) => {
```
Comment (kenodressel) [inline test/tokenCache.js:27]: can we test a positive and negative case here?

File: utils/aeternity.js [hunk 4]
```diff
@@ -97,26 +138,100 @@ class Aeternity {
...
-    return fs.readFileSync(`${__dirname}/OracleServiceInterface.aes`, 'utf-8');
-  };
+    return this.cache
+      ? this.cache.getOrSet(["getTokenRegistryState"], () => fetchData(), this.cache.shortCacheTime)
+      : fetchData();
+  }
+
+  getTokenMetaInfoCacheAccounts = async (address) => {
+    const fetchData = async () => {
+      const contract = await this.client.getContractInstance(TOKEN_CONTRACT_INTERFACE, {contractAddress: address});
```
Comment (kenodressel) [inline utils/aeternity.js:185]: do we need to re-instance the contract everytime someone looks for their balance?

File: utils/aeternity.js [hunk 5]
```diff
@@ -97,26 +138,100 @@ class Aeternity {
...
+
+      const metaInfo = await contract.methods.meta_info().then(r => r.decodedResult).catch(e => {
+        console.warn(e.message);
+        return null;
+      })
+
+      // add token to registry if not added yet
+      const tokenInRegistry = await this.getTokenRegistryState().then(state => state.find(([token, _]) => token === address));
+      if (metaInfo && !tokenInRegistry) {
+        const contract = await this.client.getContractInstance(TOKEN_REGISTRY, {contractAddress: process.env.TOKEN_REGISTRY_ADDRESS});
```
Comment (kenodressel) [inline utils/aeternity.js:195]: Could we avoid re-instantiating the registry?

File: logic/tokenCacheLogic.js (hunk 1 above)
Comment (thepiwo) [reply logic/tokenCacheLogic.js:13]: I haven't found a way not to do so, maybe you can have another look

File: test/tokenCache.js (hunk 2 above)
Comment (thepiwo) [reply test/tokenCache.js:12]: I'll check

File: utils/aeternity.js (hunk 4 above)
Comment (thepiwo) [reply utils/aeternity.js:185]: `getTokenMetaInfoCacheAccounts` itself is cached, caching the instances or keeping them is an option, but I'm not sure its worth it, I'll check

File: utils/aeternity.js (hunk 5 above)
Comment (thepiwo) [reply utils/aeternity.js:195]: I'll check

File: utils/aeternity.js [hunk 6]
```diff
@@ -182,7 +184,9 @@ class Aeternity {
 
   getTokenMetaInfoCacheAccounts = async (address) => {
     const fetchData = async () => {
-      const contract = await this.client.getContractInstance(TOKEN_CONTRACT_INTERFACE, {contractAddress: address});
+      const contract =  this.tokenContracts[address]
```
Comment (kenodressel) [inline utils/aeternity.js:187]: shouldn't you add it to the object in the else-case? or where is `this.tokenContracts` populated?

File: logic/tokenCacheLogic.js (hunk 1 above)
Comment (kenodressel) [reply logic/tokenCacheLogic.js:13]: maybe we could add a global pre-start script that initialized the SDK and the cache here https://github.com/aeternity/tipping-community-backend/blob/master/server.js#L39

File: utils/aeternity.js (hunk 6 above)
Comment (thepiwo) [reply utils/aeternity.js:187]: you are right, fixing it

File: utils/aeternity.js [hunk 7]
```diff
@@ -100,25 +143,98 @@ class Aeternity {
...
       : fetchTips();
   }
 
-  getContractSource() {
-    if (!process.env.CONTRACT_FILE) throw new Error(`env.CONTRACT_FILE is ${process.env.CONTRACT_FILE}`);
-    return fs.readFileSync(`${__dirname}/${process.env.CONTRACT_FILE}.aes`, 'utf-8');
+  async getTokenRegistryState() {
+    const fetchData = async () => this.tokenRegistry.methods.get_state().then(r => r.decodedResult);
+
+    return this.cache
```
Comment (kenodressel) [inline utils/aeternity.js:182]: are we testing the behavior of aeternity.js without cache anywhere? Is it realistic that the backend runs without cache?

Comment (thepiwo) [reply utils/aeternity.js:182]: right, this is legacy from when cache was first introduced and not mandatory to be initialized. I'll remove the checks

**Pull request closed without merging**